core/async/src/messaging.rs

```diff
@@ -50,79 +57,113 @@ impl<M> Sender<M> {
+        // Dropped error.
+        let (sender, receiver) = oneshot::channel::<Result<R, AsyncSendError>>();
+        let future = async move { receiver.await.unwrap_or_else(|_| Err(AsyncSendError::Dropped)) };
+        let responder = Box::new(move |r| sender.send(r).ok().unwrap());
+        self.send(MessageExpectingResponse { message, responder });
+        future.boxed()
+    }
+}
+
+impl<M, R: Send + 'static> Sender<MessageExpectingResponse<M, R>> {
+    /// Same as the above, but for a concrete Sender type.
+    pub fn send_async(&self, message: M) -> BoxFuture<'static, Result<R, AsyncSendError>> {
+        self.sender.send_async(message)
     }
-}
 
-/// Extension functions to wrap a CanSendAsync as an AsyncSender.
-pub trait IntoAsyncSender<M, R> {
-    /// This allows conversion of an owned CanSendAsync into an AsyncSender.
-    fn into_async_sender(self) -> AsyncSender<M, R>;
-    /// This allows conversion of a reference-counted CanSendAsync into an AsyncSender.
-    fn as_async_sender(self: &Arc<Self>) -> AsyncSender<M, R>;
+    /// Creates a sender that would handle send_async messages by producing a
+    /// result synchronously. Note that the provided function is NOT async.
+    ///
+    /// (Implementing the similar functionality with async function is possible
+    /// but requires deciding who drives the async function; a FutureSpawner
+    /// can be a good idea.)
+    pub fn from_async_fn(send_async: impl Fn(M) -> R + Send + Sync + 'static) -> Self {
+        Self::from_impl(SendAsyncFunction::new(send_async))
+    }
 }
 
-impl<M, R, T: CanSendAsync<M, R>> IntoAsyncSender<M, R> for T {
-    fn into_async_sender(self) -> AsyncSender<M, R> {
-        AsyncSender::from_impl(self)
-    }
-    fn as_async_sender(self: &Arc<Self>) -> AsyncSender<M, R> {
-        AsyncSender::from_arc(self.clone())
-    }
+/// Generic failure for async send.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum AsyncSendError {
+    // The receiver side could not accept the message.
+    Closed,
+    // The receiver side could not process the message in time.
+    Timeout,
+    // The message was ignored entirely.
+    Dropped,
 }
 
-impl<M, R> AsyncSender<M, R> {
-    pub fn send_async(&self, message: M) -> BoxFuture<'static, R> {
-        self.sender.send_async(message)
+impl Display for AsyncSendError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(self, f)
     }
+}
 
-    fn from_impl(sender: impl CanSendAsync<M, R> + 'static) -> Self {
-        Self { sender: Arc::new(sender) }
-    }
+/// Used to implement an async sender. An async sender is just a Sender whose
+/// message is `MessageExpectingResponse<M, R>`, which is a message plus a
+/// responder function (which resolves the future that send_async returns).
+pub struct MessageExpectingResponse<T, R> {
+    pub message: T,
+    pub responder: Box<dyn FnOnce(Result<R, AsyncSendError>) + Send>,
```

Comment:
I would call it `callback`. `responder` made me think that it is the entity which sends response, but from what I see it is entity which waits for response.

It is probably obvious from template args but there are already quite a lot of symbols everywhere.

The struct name is also questionable to me because I expected it simply to be `MessageWithCallback`. The message itself doesn't expect anything, it's responsibility of an user to wait for response. So the current name makes a loop in the tree of definitions and requires a reader to jump back and forth to usecases.

Reply:
Haha, rust has confusing namings, like Sender is used ubiquitously in rust libraries but it means "if you have this, you can send", whereas other languages may call it "Receiver" as in "I'm something that can receive messages that you send to me". So here, "responder" means "if you have it, you can use it to respond".

But callback is fine.